### ISSUE TITLE
Question: Understanding AttentionDecoder Implementation.

### ISSUE THREAD

**Issue by jitoledo** [farizrahman4u/seq2seq#36]
I am trying to understand the code, but it's hard for me. This is my first project using keras. 
To my understanding, the _step function can be divided in two parts:
Part1.- Compute the weighted observacion (v) 
```    
        s_tm1 = K.repeat(c_tm1, self.input_length)
        e = H + s_tm1
```
Is this a concatenation or just a regular sum? Now "e" is in the form of (nb_samples, nb_input_timesteps, hidden_dim+c_tm1_dim)?
``` 
       def a(x, states):
            output = K.dot(x, w_a) + b_a
            return output, []
        _, energy, _ = K.rnn(a, e, [], mask=None)
        energy = activations.get('linear')(energy) 
        energy = K.permute_dimensions(energy, (2, 0, 1))
        energy = energy[0]
```
Is the linear activation line really needed? It seems to return the same input. I don't understand why you permute dimmensions and then take the first element.
```  
alpha = K.softmax(energy)
```  
Can all this code from the K.rnn(a) till the K.softmax be expressed as a TimeDistributedDense(activation='softmax') over the sequence "e"?
```    
        alpha = K.repeat(alpha, self.hidden_dim)
        alpha = K.permute_dimensions(alpha, (0, 2 , 1))
        weighted_H = H * alpha
        v = K.sum(weighted_H, axis=1)
```
Part 2.- Feed as that V as an input it to a standard LSTM (The code seems like the standard implementation of the LSTM, am I right ?)

**Comment by farizrahman4u** [comment]
Hi!
Great questions. Let me answer them one by one.

> Is this a concatenation or just a regular sum? Now "e" is in the form of (nb_samples, nb_input_timesteps, hidden_dim+c_tm1_dim)?

Its a sum. `+` means sum. Both `H` and `s_tm1` are of same shape, so when we add them, we get a sum with no change in the dimensions. Nothing fancy here. Its element-wise addition. ```c = a + b => c[i][j][k] = a[i][j][k] + b[i][j][k] for all i, j, k```.

> Is the linear activation line really needed? 

No, that line has no effect. I was actually experimenting with different activation functions.. and ended up with linear.

> I don't understand why you permute dimmensions and then take the first element.

The RNN returns scalars for each timesteps, so the shape of energy would be `(samples, timesteps, 1)`. I am simply removing the `1` dimension. It can be done more elegantly using the `squeeze` function. This will be fixed soon.

> Can all this code from the K.rnn(a) till the K.softmax be expressed as a TimeDistributedDense(activation='softmax') over the sequence "e"?

Not exactly, because there is some reshaping involved. But the general idea is the same. The code from `def a(x, states):` upto  `_, energy, _ = K.rnn` can abstracted into a `TimeDistributedDense` layer. You can not include the `softmax` step in the `TimeDistributedDense` layer because a `TimeDistributedDense` with `activation='softmax'` will do the softmax **on each** timestep, instead of doing a softmax **across** the timesteps.

>  Feed as that V as an input it to a standard LSTM (The code seems like the standard implementation of the LSTM, am I right ?)

Yes!
